domain/removal/state/model/unit.go

```diff
@@ -70,28 +71,77 @@ AND    life_id = 0`, unitUUID)
 		return cascaded, errors.Errorf("preparing unit life update: %w", err)
 	}
 
-	var mUUID string
-	if err := errors.Capture(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+	var (
+		mUUID        string
+		sAttachments []string
+		sInstances   []string
+	)
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		if err := tx.Query(ctx, updateUnitStmt, unitUUID).Run(); err != nil {
 			return errors.Errorf("advancing unit life: %w", err)
 		}
 
+		sAttachments, err = st.ensureUnitStorageAttachmentsNotAlive(ctx, tx, uUUID)
+		if err != nil {
+			return errors.Errorf("setting unit storage attachment lives to dying: %w", err)
+		}
+
 		mUUID, err = st.markMachineAsDyingIfAllUnitsAreNotAlive(ctx, tx, uUUID)
 		if err != nil {
-			return errors.Errorf("marking last unit on machine as dying: %w", err)
+			return errors.Errorf("setting unit machine life to dying: %w", err)
 		}
 
 		return nil
-	})); err != nil {
-		return cascaded, err
+	}); err != nil {
+		return cascaded, errors.Capture(err)
 	}
 
 	if mUUID != "" {
 		cascaded.MachineUUID = &mUUID
 	}
+	cascaded.StorageAttachmentUUIDs = sAttachments
+	cascaded.StorageInstanceUUIDs = sInstances
 	return cascaded, nil
 }
 
+func (st *State) ensureUnitStorageAttachmentsNotAlive(
+	ctx context.Context, tx *sqlair.TX, uUUID string,
+) ([]string, error) {
+	type attachmentUUID entityUUID
+	unitUUID := entityUUID{UUID: uUUID}
+
+	stmt, err := st.Prepare(`
+SELECT uuid AS &attachmentUUID.*
+FROM   storage_attachment
+WHERE  unit_uuid = $entityUUID.uuid
+AND    life_id = 0`, unitUUID, attachmentUUID{})
```

Comment:
I'm guessing we're doing this for clarity, as it isn't required.

```suggestion
	unitUUID := entityUUID{UUID: uUUID}

	stmt, err := st.Prepare(`
SELECT uuid AS &entityUUID.*
FROM   storage_attachment
WHERE  unit_uuid = $entityUUID.uuid
AND    life_id = 0`, unitUUID)
```